lib/oneFFS.ml

```diff
@@ -163,3 +163,50 @@ module Make(B : Mirage_block.S) = struct
       Cstruct.blit_from_string Header.empty 0 buf 0 (String.length Header.empty);
       Lwt.return { b; info; f; empty_header = buf; }
 end
+
+module Make_lossy = Make
+
+module Make_lossless(Pclock : Mirage_clock.PCLOCK)(Block : Mirage_block.S) = struct
+  module FS = Filesystem.Make(Pclock)(Block)
+
+  type t = {
+    b : Block.t;
+    mutable superblock : FS.superblock;
+  }
+
+  let connect b =
+    let* info = Block.get_info b in
+    if info.sector_size < FS.superblock_size then
+      failwith "sector size too small for this filesystem";
+    let* r = FS.read_data b in
+    match r with
+    | Error `Bad_checksum ->
```

Comment:
not 100% sure this behaviour is intended... so we read something bad from the disk... but does that mean we want to initialize and return "we have read nothing"? if a user has a bad disk with broken blocks, shouldn't that at least be mentioned in a log message that the checksum verification failed (so the user can try to find a backup)?

Reply:
Hm, yes. I will add a log message.

FWIW `FS.read_data` will try to fix a broken superblock if the other one passes checksum. I took this code from tlstunnel. It seems `` `Bad_checksum `` is returned also on an uninitialized disk - so the intent was maybe to initialize the disk if it wasn't? I can add a separate `format` function to do that.

Reply:
That's right, it is as well used for formatting... I think a log message is fine -- the API client receives `None` as value, so if they expected data they should do something about it ;)